microprofile/connectors/kafka/src/main/java/io/helidon/microprofile/connectors/kafka/BasicKafkaConsumer.java

```diff
@@ -0,0 +1,206 @@
+        if (!partitionsAssignedLatch.await(timeout, unit)) {
+            throw new TimeoutException("Timeout for subscription reached");
+        }
+    }
+
+    /**
+     * Close gracefully. Stops wakes possible blocked poll and close consumer.
+     */
+    @Override
+    public void close() {
+        // Stops pooling
+        consumer.wakeup();
+        // Wait that current task finishes in case it is still running
+        taskLock.lock();
+        taskLock.unlock();
+        LOGGER.fine("Closing kafka consumer");
+        consumer.close();
+    }
+
+    //Move to messaging incoming connector
+    private void runInNewContext(Runnable runnable) {
+        Context parentContext = Context.create();
+        Context context = Context
+                .builder()
+                .parent(parentContext)
+                .id(String.format("%s:message-%s", parentContext.id(), UUID.randomUUID().toString()))
+                .build();
+        Contexts.runInContext(context, runnable);
+    }
+
+    private final class BackPressureLayer implements Runnable {
+
+        private final LinkedList<ConsumerRecord<K, V>> backPressureBuffer = new LinkedList<>();
+        private final LinkedList<CompletableFuture<Void>> ackFutures = new LinkedList<>();
+        private final Subscriber<? super KafkaMessage<K, V>> subscriber;
+        private final long poolTimeout;
+
+        private BackPressureLayer(Subscriber<? super KafkaMessage<K, V>> subscriber, long poolTimeout) {
+            this.subscriber = subscriber;
+            this.poolTimeout = poolTimeout;
+        }
+
+        @Override
+        public void run() {
+            try {
+                taskLock.lock();
+                waitForAcksAndPoll();
+                ConsumerRecord<K, V> cr;
+                while ((cr = backPressureBuffer.poll()) != null) {
+                    KafkaMessage<K, V> kafkaMessage = new KafkaMessage<>(cr);
+                    ackFutures.add(kafkaMessage.getAckFuture());
+                    runInNewContext(() -> subscriber.onNext(kafkaMessage));
+                }
+            } finally {
+                taskLock.unlock();
+            }
+        }
+
+        /**
+         * Naive impl of back pressure wise lazy poll.
+         * Wait for the last batch of records to be acknowledged before commit and another poll.
+         */
+        private void waitForAcksAndPoll() {
+            if (backPressureBuffer.isEmpty()) {
+                try {
+                    if (!ackFutures.isEmpty()) {
+                        LOGGER.fine(String.format("Wait for %s ACKs", ackFutures.size()));
+                        CompletableFuture.allOf(ackFutures.toArray(new CompletableFuture[0])).get();
+                        ackFutures.clear();
+                        consumer.commitSync();
+                    }
+                    consumer.poll(Duration.ofMillis(poolTimeout)).forEach(backPressureBuffer::add);
```

Comment:
Definitely `pollTimeout` not `poolTimeout`

Reply:
Done